pageserver/src/tenant/storage_layer/inmemory_layer.rs

```diff
@@ -76,6 +68,10 @@ pub struct InMemoryLayerInner {
     /// Each serialized Value is preceded by a 'u32' length field.
     /// PerSeg::page_versions map stores offsets into this file.
     file: EphemeralFile,
+
+    /// A buffer for serializing object during [`InMemoryLayer::put_value`].
+    /// This buffer is reused for each serialization to avoid additional malloc calls.
+    ser_buffer: Vec<u8>,
```

Comment:
Potentially risky change, because instead of fixed amount of buffers stuck to threads around we have `N` (=timelines). I don't think this could be [`task_local`](https://docs.rs/tokio/latest/tokio/macro.task_local.html) because we should have the same amount of those tasks.

Is the next to step make the `EphmeralFile::write_blob::Writer` into `AsyncWrite`? I don't think there has every been a need to buffer these, but making a TLS buffer is too convenient. Or does the `write_blob` need to be memcpy fast, or did it need to be for some synchronization issues? Would imagine not, the action is limited to single cell in the pagecache.

Reply:
To explain, there's double-buffering with the write_blob::Writer and this ser_buffer, it's unncessary.

> Or does the `write_blob` need to be memcpy fast

Of course it does not need to be because we are doing the small memcpies already to `ser_buffer`, so that we can do the fewer large memcpies to page cache buffer.

Reply:
75MB per 10k timelines at 8k. We could probably afford this as is. Should we prealloc to 8kB or could we have larger..?

Reply:
If it's bounded by 8 KB we could also just have that on the stack instead, i.e. through SmallVec.

Reply:
Mhh thinking of it, this seems like one could either use tokio's `task_local` macro (although I'm not sure if even that one supports access over await points), or one could just use `smallvec` with a reasonable default (like 128 bytes). If a value is above 128 bytes, this should happen rarely, we can afford an allocation.

It's important to note that anything stored on the stack by a task will be stored somewhere, so it contributes to memory overhead just as easily.

Reply:
> Is the next to step make the `EphmeralFile::write_blob::Writer` into `AsyncWrite`?

No, this is not my plan. The `Writer` is just a struct. There is no trait involvement, and I would like to keep it that way. My plan is to make the page cache access in `push_bytes` async, and for that `push_bytes` needs to be async as well.

Reply:
Hm, would be nice if the be_coder() could be an `AsyncRead`

Reply:
Bincode 1.0 has no async serialization, and it looks like this is also not going to come for bincode 2: https://github.com/bincode-org/bincode/issues/519#issuecomment-1061925868

Reply:
The change is now to use SmallVec of 128 bytes. Did you test how large writes happen to these in our test suite? Well, I guess `pageserver::walrecord` the unpacked ones could fit into that nicely.

Reply:
I am good to resolve this.

Reply:
The over 14k is interesting. Do you have idea what that was?